### ISSUE TITLE
Support Entity-to-Entity Scheduled SignalEntities for OutOfProc PLs

### ISSUE THREAD

**Issue by davidmrdavid** [Azure/azure-functions-durable-extension#1554]
Currently, the `signalEntity` operation of the entity context API does not support specifying a firing time in our out-of-process offerings. The reason for this limitation, I believe, is that the extension does not provide a "scheduledTimeUTC" field for the `Signal` `struct`, which ultimately means that an out-of-proc PL does not have any way to communicate a non-immediate signal.

Below is the **current** definition of the `Signal` `struct`, for reference:

https://github.com/Azure/azure-functions-durable-extension/blob/3bdd5704f8ac41b4c549d6e0d811a6428d9c4c1f/src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs#L583-L605

In addition to adding another field in which to store the signal's firing time, we need to alter our event-processing logic to use that field in order to determine when the signal is to be fired. It's unclear to me how best to do that at the moment.

I do see that we have some scheduling logic below:

https://github.com/Azure/azure-functions-durable-extension/blob/f95755b5bdba8e3a0837da013ca9dcbeb7b611c9/src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs#L682-L724

However, that scheduling logic uses the `scheduledTimeUTC` field of a `RequestMessage` object, not of a `Signal` struct. So either we'll have to add special logic for out-of-proc PLs, or we'll have to figure out a way to propagate the `Signal` field up to a `RequestMessage` containing it. I'm not exactly sure how to do that at the moment.

However, the block of code embedded below may have something to do with it:

https://github.com/Azure/azure-functions-durable-extension/blob/3bdd5704f8ac41b4c549d6e0d811a6428d9c4c1f/src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableEntityContext.cs#L330-L363

Folks, any suggestions on how best to approach this are greatly appreciated. @ConnorMcMahon , can we add this as a goal for our upcoming Durable-Extension release? _Thanks!_

**Comment by cgillum** [comment]
In addition to updating our out-of-proc schema, I was thinking we'd just make a small change to OutOfProcOrchestrationShim.cs to support this. Specifically, [this line](https://github.com/Azure/azure-functions-durable-extension/blob/f1a429106980eb8b824b5a0ce6a7cbc0ac0193ad/src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs#L168).

![image](https://user-images.githubusercontent.com/2704139/98616799-650ff600-22b2-11eb-9a3d-67c503e195ba.png)

Notice there is an overload that takes the `scheduledTimeUtc` as you pointed out in your last code snippet above. Can we just call the other `SignalEntity` overload depending on whether the action we get contains this new timestamp?

**Comment by davidmrdavid** [comment]
I think the change you suggest would only work for the orchestrator-to-entity signalEntity API, but not for entity-to-entity signals, right? My understanding was that the `OutOfProcOrchestrationShim` is only "hit" by orchestrators, but I could be wrong about that.

For context, this ticket is inspired by this issue: https://github.com/Azure/azure-functions-durable-js/issues/140#issuecomment-724397148
which specifically requests entity-to-entity scheduled signals for JS.

**Comment by ConnorMcMahon** [comment]
@davidmrdavid, unless you think you can squeeze this into 2.4.0 (i.e. knock this PR out in a couple of hours),  let's put this on the backlog and we can prioritize it for the following release.

**Comment by davidmrdavid** [comment]
Yup, I have not had the time to prioritize this yet. I'll put it on the backlog

**Comment by YunchuWang** [comment]
Closing for the .NET isolated worker scope. Entity-to-entity scheduled signals are supported starting with Microsoft.Azure.Functions.Worker.Extensions.DurableTask 1.4.0 (Microsoft.DurableTask.* 1.11.0) through SignalEntityOptions.SignalTime and the isolated gRPC scheduledTime field. This resolution does not include .NET in-process or the legacy JSON out-of-process entity shim.
